### PR TITLE
Added .DEFINEDMACRO psuedo function

### PR DESCRIPTION
**Pull request by Movax12**
This is a separate branch from the other work I was doing and can be evaluated on it's own.

This allows ca65 to check if an identifier is a macro. This is actually needed by me to be able to call other macros cleanly from within macros. (I need to evaluate if a parameter is a macro, instruction or variable.)

### CODE REVIEW

File: src/ca65/expr.c [hunk 1]
```diff
@@ -417,6 +418,33 @@ static ExprNode* FuncDefined (void)
...
+            Mac = FindMacro (&CurTok.SVal);
+        } else {
+            /* Macros and symbols may NOT use the names of instructions */
+            if (FindInstruction (&CurTok.SVal) < 0) {
+                Mac = FindMacro (&CurTok.SVal);
+            }
+        }
+        NextTok();
+    } else {
+        Error("Idenitifier expected.");
```
Comment (greg-king5) [inline src/ca65/expr.c]: `Identifier` typo.

File: src/ca65/expr.c [hunk 2]
```diff
@@ -417,6 +418,33 @@ static ExprNode* FuncDefined (void)
...
+        if (UbiquitousIdents) {
+            /* Macros CAN be instructions, so check for them first */
+            Mac = FindMacro (&CurTok.SVal);
+        } else {
+            /* Macros and symbols may NOT use the names of instructions */
+            if (FindInstruction (&CurTok.SVal) < 0) {
+                Mac = FindMacro (&CurTok.SVal);
+            }
+        }
+        NextTok();
```
Comment (oliverschmidt) [inline src/ca65/expr.c]: Missing space.

File: src/ca65/expr.c (hunk 1 above)
Comment (oliverschmidt) [inline src/ca65/expr.c]: Missing space.

File: src/ca65/expr.c [hunk 3]
```diff
@@ -417,6 +418,33 @@ static ExprNode* FuncDefined (void)
...
+    } else {
+        Error("Idenitifier expected.");
+    }
+
+    return GenLiteralExpr(Mac != 0);
```
Comment (oliverschmidt) [inline src/ca65/expr.c]: Missing space.

File: src/ca65/expr.c [hunk 4]
```diff
@@ -417,6 +418,33 @@ static ExprNode* FuncDefined (void)
...
+
+    /* Check for a macro or an instruction depending on UbiquitousIdents */
+
+    if (CurTok.Tok == TOK_IDENT) {
+        if (UbiquitousIdents) {
+            /* Macros CAN be instructions, so check for them first */
+            Mac = FindMacro (&CurTok.SVal);
+        } else {
+            /* Macros and symbols may NOT use the names of instructions */
+            if (FindInstruction (&CurTok.SVal) < 0) {
```
Comment (greg-king5) [inline src/ca65/expr.c]: The `UbiquitousIdents` test isn't needed.  Macroes with instruction names are blocked from being created when `UbiquitousIdents` is false.  Therefore, `FindMacro()` would return `NULL` when it's given a mnemonic.

File: src/ca65/expr.c [hunk 5]
```diff
@@ -417,6 +418,33 @@ static ExprNode* FuncDefined (void)
...
+                Mac = FindMacro (&CurTok.SVal);
+            }
+        }
+        NextTok ();
```
Comment (greg-king5) [inline src/ca65/expr.c]: It would be nice to have the "usual" comment above here:
`/* Skip the name */`

**Comment by greg-king5** [comment]
This looks like a safe patch.  It doesn't conflict with anything.  And, the patch is structured like all of the other Assembly dot-function code.  I think that you don't need to consider it to be experimental.  It doesn't need to be an optional feature.  It should be a normal part of ca65's Assembly Language.

File: src/ca65/expr.c (hunk 4 above)
Comment (Movax12) [inline src/ca65/expr.c]: You are probably right. I was thinking about that actually. I'll verify with some testing.

**Comment by Movax12** [comment]
I'm happy to not have it experimental if there is no need for it to be. If there are no objections I'll remove the code and documentation for `.feature`

**Comment by oliverschmidt** [comment]
Hi,

I'm by now pretty sure we have a communication deadlock here :-(

@Movax12: You should know that GitHub doesn't send notifications when you push new commits to your pull request branch. So if you feel that you've taken care of the comments/feedback on the pull request please post a comment here, even if it just says "Updated".

@greg-king5: If you are aware of the latest commits in this pull request and feel that it is ready to be merged from your perspective than it would be nice if you would state so in a comment. I admit that I don't closely follow all aspects of the discussion here so I can't judge easily if all open "issues" have been clarified or taken care of.

I'm sorry if one or both of you should feel that I'm not taking care enough of these pull requests but I'm after all no active user of any of the advanced ca65 features and therefore I'm having a hard time to judge if a feature
* makes sense at all
* is implemented correctly

Regards,
Oliver

**Comment by oliverschmidt** [comment]
Please rebase this pull request and I'll merge it.

File: src/ca65/scanner.c [hunk 6]
```diff
@@ -132,163 +132,164 @@ struct DotKeyword {
...
+    { ".CONSTRUCTOR",   TOK_CONSTRUCTOR  },
+    { ".CPU",           TOK_CPU          },
+    { ".DATA",          TOK_DATA         },
+    { ".DBG",           TOK_DBG          },
+    { ".DBYT",          TOK_DBYT         },
+    { ".DEBUGINFO",     TOK_DEBUGINFO    },
+    { ".DEF",           TOK_DEFINED      },
+    { ".DEFINE",        TOK_DEFINE       },
+    { ".DEFINED",       TOK_DEFINED      },
+    { ".DEFINEDMACRO",  TOK_DEFINEDMACRO },
```
Comment (greg-king5) [inline src/ca65/scanner.c]: I suggest that, after the rebase, you move the right edge (the column of right braces) of this initiator over to the next 8-character tab stop (column 49).  Then, if anyone adds a longer token to this list, the other lines might not need to be changed.

**Comment by Movax12** [comment]
I moved .ismnemonic code and documentation to make it alphabetical.

**Comment by greg-king5** [comment]
Am I correct to assume that this PR is -- again -- ready to be merged?

**Comment by Movax12** [comment]
Yes - This is my first rebase and first time merging conflicts manually though, so take a close look at it.

**Comment by greg-king5** [comment]
You might want to move the FuncIsMnemonic() function down to a spot just below FuncHiWord().  But, it's only a very small style issue -- the rest of the rebase looks great!

**Comment by oliverschmidt** [comment]
@Movax12: Are you going to move the function definition as suggested or do you want me to merge as-is?

**Comment by Movax12** [comment]
I'll move it - hopefully late today.

**Comment by Movax12** [comment]
Done .. merge at will.

**Comment by oliverschmidt** [comment]
Thanks :-)

**Pull request merged**